parsec/core/gui/claim_user_widget.py

```diff
@@ -1,207 +1,660 @@
+from PyQt5.QtCore import pyqtSignal
+from PyQt5.QtGui import QFont
+from PyQt5.QtWidgets import QWidget, QDialog
+
+from enum import IntEnum
+
+import trio
 
 from structlog import get_logger
 
-from parsec.api.protocol import DeviceID
-from parsec.core.types import BackendOrganizationClaimUserAddr
+from parsec.api.protocol import DeviceID, HumanHandle
+from parsec.core.types import LocalDevice
 from parsec.core.local_device import LocalDeviceAlreadyExistsError, save_device_with_password
-from parsec.core.invite_claim import (
-    claim_user as core_claim_user,
-    InviteClaimError,
-    InviteClaimBackendOfflineError,
-)
+
+from parsec.core.invite import claimer_retrieve_info
+from parsec.core.backend_connection import backend_invited_cmds_factory, BackendConnectionRefused
+
 from parsec.core.gui import validators
 from parsec.core.gui.trio_thread import JobResultError, ThreadSafeQtSignal
 from parsec.core.gui.desktop import get_default_device
 from parsec.core.gui.custom_dialogs import show_error, GreyedDialog, show_info
 from parsec.core.gui.lang import translate as _
-from parsec.core.gui.password_validation import PasswordStrengthWidget, get_password_strength
+from parsec.core.gui.password_validation import get_password_strength
 from parsec.core.gui.ui.claim_user_widget import Ui_ClaimUserWidget
+from parsec.core.gui.ui.claim_user_code_exchange_widget import Ui_ClaimUserCodeExchangeWidget
+from parsec.core.gui.ui.claim_user_provide_info_widget import Ui_ClaimUserProvideInfoWidget
+from parsec.core.gui.ui.claim_user_instructions_widget import Ui_ClaimUserInstructionsWidget
+from parsec.core.gui.ui.claim_user_finalize_widget import Ui_ClaimUserFinalizeWidget
 
 logger = get_logger()
 
 
-async def _do_claim_user(
-    config,
-    password: str,
-    password_check: str,
-    token: str,
-    user_id: str,
-    device_name: str,
-    organization_addr: BackendOrganizationClaimUserAddr,
-):
-    if password != password_check:
-        raise JobResultError("password-mismatch")
-    if len(password) < 8:
-        raise JobResultError("password-size")
-
-    try:
-        device_id = DeviceID(f"{user_id}@{device_name}")
-    except ValueError as exc:
-        raise JobResultError("bad-device_name") from exc
-
-    try:
-        device = await core_claim_user(
-            organization_addr=organization_addr.to_organization_addr(),
-            new_device_id=device_id,
-            token=token,
-            keepalive=config.backend_connection_keepalive,
-        )
+class Claimer:
+    class Step(IntEnum):
+        RetrieveInfo = 1
+        WaitPeer = 2
+        GetGreeterSas = 3
+        SignifyTrust = 4
+        GetClaimerSas = 5
+        WaitPeerTrust = 6
+        ClaimUser = 7
 
-    except InviteClaimBackendOfflineError as exc:
-        raise JobResultError("backend-offline", info=str(exc)) from exc
+    def __init__(self):
+        self.main_oob_send, self.main_oob_recv = trio.open_memory_channel(0)
```

Comment:
Not sure about the naming: "oob" stands for "Out of bound" (which refer to the fact something is passed outside of the application)...

I think it would be better to enforce a naming like "worker" for the coroutine actually doing the work and "driver" for the coroutines requesting operations (would be cool to add a bit of documentation about this in the Claimer docstring to easily get the basic concepts of the module)

So we could name them `self.worker_send_channel`, `self.worker_receive_channel`, `self.driver_send_channel`, `self.driver_receive_channel`.

Reply:
Another consideration is the driver channel that is shared between multiple producers (i.e. multiple Widgets that will trigger multiple jobs). This can cause issues if a job get cancelled before it can retrieve the answer to his request, hence the next job will get it instead of the answer it was expecting...

The cleanest solution for this would be to only keep the driver channel, and send in this channel a reply channel in addition with the command to execute.

However in case something goes wrong (and a job got cancelled), the channel should not be reused given the widget reset recreate Claimer. I wonder if we could take advantage of that to the [close system in trio channel](https://trio.readthedocs.io/en/stable/reference-core.html#trio.open_memory_channel) to ensure we cannot end up in a subtle weird state.